pkg/services/ngalert/schedule/schedule.go

```diff
@@ -289,6 +284,17 @@ func (sch *schedule) processTick(ctx context.Context, dispatcherGroup *errgroup.
 				rule:        item,
 				folderTitle: folderTitle,
 			}})
+		} else if !rulesDiff.IsEmpty() {
```

Comment:
nit: Personally, I find `else if` so much harder to reader. What do you think of my alternative?

```go
// If it's not time to evaluate the ruler yet but the rule version has been updated, let's notify the rule routine.
if !ruleReadyToRun && !rulesDiff.IsEmpty() {
	_, ok := rulesDiff.updated[key]
	if ok {
		sch.log.Debug("Rule has been updated. Notifying evaluation routine", key.LogContext()...)
		go func() {
			ruleInfo.update(ruleVersionAndPauseStatus{Version: ruleVersion(item.Version), IsPaused: item.IsPaused})
		}()
	}
}
```

This does two things:

- There's no reason for you to build and allocate the whole `needToUpdate`, which would notify all the routines at the rules at the same time, when you can do this incrementally (unlike readyToRun) there's no need to wait.
- The introduction of `ruleReadyToRun` makes the code a bit simpler to read (subjective I know) at the expense of having to declare `var ruleReadyToRun bool` and then setting  `ruleReadyToRun = true` when you add something to `readyToRun`.

Reply:
With this, you don't need to modify the signature of `processTick`